dot/core/service.go

```diff
@@ -187,7 +187,14 @@ func (s *Service) handleBlock(block *types.Block, state *rtstorage.TrieState) er
 		block.Header.Hash(), state.MustRoot())
 
 	rt, err := s.blockState.GetRuntime(&block.Header.ParentHash)
-	if err != nil {
+	if errors.Is(err, blocktree.ErrFailedToGetRuntime) {
+		rt, err = s.getRuntimeFromDB(&block.Header.ParentHash)
+		if err != nil {
+			return err
+		}
+		defer rt.Stop()
```

Comment:
nit maybe let's add a comment why this is needed

```suggestion
		// ensure the runtime stops and releases resources since it was 
		// instantiated from disk just for this function call.
		defer rt.Stop()
```